src/cls/otp/cls_otp.cc

```diff
@@ -0,0 +1,510 @@
+    ENCODE_START(1, 1, bl);
+    ::encode(otp, bl);
+    ::encode(last_checks, bl);
+    ::encode(last_success, bl);
+    ENCODE_FINISH(bl);
+  }
+  void decode(bufferlist::iterator &bl) {
+    DECODE_START(1, bl);
+    ::decode(otp, bl);
+    ::decode(last_checks, bl);
+    ::decode(last_success, bl);
+    DECODE_FINISH(bl);
+  }
+
+  void trim_expired(const ceph::real_time& now);
+  void check(const string& token, const string& val, bool *update);
+  bool verify(const ceph::real_time& timestamp, const string& val);
+
+  void find(const string& token, otp_check_t *result);
+};
+WRITE_CLASS_ENCODER(otp_instance)
+
+
+void otp_instance::trim_expired(const ceph::real_time& now)
+{
+  ceph::real_time window_start = now - make_timespan(otp.step_size);
+
+  while (!last_checks.empty() &&
+         last_checks.front().timestamp < window_start) {
+    last_checks.pop_front();
+  }
+}
+
+void otp_instance::check(const string& token, const string& val, bool *update)
+{
+  ceph::real_time now = ceph::real_clock::now();
+  trim_expired(now);
+
+  if (last_checks.size() >= ATTEMPTS_PER_WINDOW) {
+    /* too many attempts */
+    *update = false;
+    return;
+  }
+
+  otp_check_t check;
+  check.token = token;
+  check.timestamp = now;
+  check.result = (verify(now, val) ? OTP_CHECK_SUCCESS : OTP_CHECK_FAIL);
+
+  last_checks.push_back(check);
+
+  *update = true;
+}
+
+bool otp_instance::verify(const ceph::real_time& timestamp, const string& val)
+{
+  size_t slen = otp.seed.size() / 2 + 1;
+  char secret[slen];
+  uint64_t index;
+  uint32_t secs = (uint32_t)ceph::real_clock::to_time_t(timestamp);
+  int result = oath_hex2bin(otp.seed.c_str(), secret, &slen);
+  if (result != OATH_OK) {
+    CLS_LOG(20, "failed to parse seed");
+    return result;
+  }
+
+  result = oath_totp_validate3(secret, slen,
+                                   secs, otp.step_size, 0, otp.window,
+                                   nullptr /* otp pos */, &index,
+                                   val.c_str());
+  if (result == OATH_INVALID_OTP ||
+      result < 0) {
+    CLS_LOG(20, "otp check failed, result=%d", result);
+    return false;
+  }
+
+  if (index <= last_success) { /* already used value */
+    CLS_LOG(20, "otp, use of old token: index=%lld last_success=%lld", (long long)index, (long long)last_success);
+    return false;
+  }
+
+  last_success = index;
+
+  return true;
+}
+
+void otp_instance::find(const string& token, otp_check_t *result)
+{
+  ceph::real_time now = real_clock::now();
+  trim_expired(now);
+
+  for (auto entry : boost::adaptors::reverse(last_checks)) {
+    if (entry.token == token) {
+      *result = entry;
+      return;
+    }
+  }
+  result->token = token;
+  result->result = OTP_CHECK_UNKNOWN;
+  result->timestamp = now;
+}
+
+static int get_otp_instance(cls_method_context_t hctx, const string& id, otp_instance *instance)
+{
+  bufferlist bl;
+  string key = otp_key_prefix + id;
+ 
+  int r = cls_cxx_map_get_val(hctx, key.c_str(), &bl);
```

Comment:
the `cls_cxx_` apis all take `std::string` - so these are converting to `const char*` and back again